server/src/main/java/org/elasticsearch/common/util/CollectionUtils.java

```diff
@@ -427,4 +444,41 @@ public static int sortAndDedup(final BytesRefArray bytes, final int[] indices) {
 
         return result;
     }
+
+    /**
+     * Creates a {@code HashMap} instance, with a high enough "initial capacity" that it <i>should</i> hold
+     * {@code expectedSize} elements without growth. This behavior cannot be broadly guaranteed, but it is observed
+     * to be true for OpenJDK 1.7.
+     * It also can't be guaranteed that the method isn't inadvertently <i>oversizing</i> the returned map.
+     *
+     * <p>From Guava 24.1-jre. <b>com.google.common.collect.Maps </b></p>
+     *
+     * @param expectedSize the number of entries you expect to add to the returned map
+     * @return a new, empty {@code HashMap} with enough capacity to hold {@code expectedSize} entries without resizing
+     * @throws IllegalArgumentException if {@code expectedSize} is negative
+     */
+    public static <K, V> HashMap<K, V> newHashMapWithExpectedSize(int expectedSize) {
+        return new HashMap<>(capacity(expectedSize));
+    }
+
+    /**
+     * Returns a capacity that is sufficient to keep the map from being resized as long as it grows no larger
+     * than expectedSize and the load factor is ≥ its default (0.75).
+     *
+     * <p>From Guava 24.1-jre. <b>com.google.common.collect.Maps </b></p>
+     */
+    static int capacity(int expectedSize) {
+        if (expectedSize < 3) {
+            if (expectedSize < 0) {
+                throw new IllegalArgumentException("expectedSize cannot be negative but was: " + expectedSize);
+            }
+            return expectedSize + 1;
+        }
+        if (expectedSize < MAX_POWER_OF_TWO) {
+            // This is the calculation used in JDK8,9,10 to resize when a putAll happens.
+            // It seems to be the most conservative calculation we can make. 0.75 is the default load factor.
+            return (int) ((float) expectedSize / 0.75F + 1.0F);
+        }
+        return Integer.MAX_VALUE; // any large value
+    }
```

Comment:
I would refrain from adding these to be honest, especially if only used in our client/parsing code.

Reply:
++

For the most part we're ok with `new HashMap<>(expectedSize);`